src/Equations/Fluid/ComptonRateTerm.cpp

```diff
@@ -0,0 +1,109 @@
+#include "DREAM/Equations/Fluid/ComptonRateTerm.hpp"
+#include "DREAM/NotImplementedException.hpp"
+#include "DREAM/DREAMException.hpp"
+
+/**
+ * Implementation of a class which represents the Gamma_compton contribution to the n_re equation.
+ * Employs the analytical growth rate calculated by RunawayFluid.
+ */
+using namespace DREAM;
+/**
+ * Constructor.
+ */
+ComptonRateTerm::ComptonRateTerm(
+    FVM::Grid *g, FVM::UnknownQuantityHandler *uqn,
+    RunawayFluid *rf, real_t scaleFactor
+) : EquationTerm(g), REFluid(rf), scaleFactor(scaleFactor) {
+
+    this->AllocateGamma();
+
+    this->id_n_tot   = uqn->GetUnknownID(OptionConstants::UQTY_N_TOT);
+}
+
+/**
+ * Destructor.
+ */
+ComptonRateTerm::~ComptonRateTerm() {
+    DeallocateGamma();
+}
+
+/**
+ * Allocate memory for the runaway rate.
+ */
+void ComptonRateTerm::AllocateGamma() {
+    this->gamma = new real_t[this->grid->GetNr()];
+}
+
+/**
+ * Free memory for the runaway rate.
+ */
+void ComptonRateTerm::DeallocateGamma() {
+    delete [] this->gamma;
+}
+
+/**
+ * Method called when the grid has been rebuilt.
+ */
+bool ComptonRateTerm::GridRebuilt() {
+    DeallocateGamma();
+    AllocateGamma();
+
+    return true;
+}
+
+void ComptonRateTerm::Rebuild(const real_t, const real_t, FVM::UnknownQuantityHandler *uqn) {
+    const len_t nr = this->grid->GetNr();
+
+    for (len_t ir = 0; ir < nr; ir++)
+        this->gamma[ir] = REFluid->GetComptonRunawayRate(ir);
+
+    this->data_n_tot   = uqn->GetUnknownData(id_n_tot);
+}
+ 
+/**
+ * Set jacobian block for this term. As for the avalanche growthrate, we neglect derivatives of pstar
+ * so that the compton source is treated as a constant times the total electron density
+ */
+void ComptonRateTerm::SetJacobianBlock(const len_t, const len_t derivId, FVM::Matrix *jac, const real_t*){
+    const len_t nr  = this->grid->GetNr();
+    len_t offset = 0;
+
+    for(len_t ir=0;ir<nr;ir++){
+        const len_t np1 = this->grid->GetMomentumGrid(ir)->GetNp1();
+        const len_t np2 = this->grid->GetMomentumGrid(ir)->GetNp2();
+
+        if(derivId==id_n_tot){
```

Comment:
It's probably better to put this if statement outside of the for loop for now. This saves a few branch checks (which should generally be avoided within loops)